### DISCUSSION TITLE
zkSpin - Building a Decentralized Future for Privacy Games with Noir and MPC on Aztec

### DISCUSSION THREAD

**Discussion by wadesonghang** [noir-lang/noir#6336] [NRG#2 Private Shared States]
# **Summary**

We aim to develop a decentralized, privacy-preserving poker game on the **Aztec** using **Noir** and **Multi-Party Computation (MPC)**. The primary goal is to maintain the privacy of players' hands while ensuring transparent gameplay for actions such as betting and determining winners. Noir enables ZKPs to verify player actions without revealing private data. Additionally, MPC is used for fair card shuffling, ensuring randomness without a trusted dealer. The game will use Aztec’s smart contracts for private transactions and zkRollups for scaling, enabling secure and private poker gameplay on a decentralized platform.

# **Methodology**

## Research approach:

The approach is based on the cryptographic foundations of **mental poker** to enable a decentralized, secure poker game. It is a well-known concept in cryptography, originating from the need to simulate the randomness and secrecy of physical card games without requiring a trusted dealer. Traditionally, these protocols enable players to jointly shuffle and deal cards in a way that ensures privacy without revealing any hands prematurely.

In this implementation, **Noir** serves as the backbone for handling the cryptographic proofs, specifically ZKP, to maintain player privacy while ensuring the integrity of gameplay. The core of the game architecture rests on the following principles:

- **Noir for ZKPs**: Each player’s actions during the game, such as betting or folding, are validated via ZKPs generated using **Noir**. This ensures that while the actual data (such as the cards in hand) remains private, the system can still verify that a player’s actions are legitimate. Each player’s cards and private actions remain confidential, but their public actions (e.g., bet size) are recorded on-chain.
- **MPC for shuffling and dealing**: CCard shuffling and dealing will be handled through a committee who using TAECO MPC protocol. Each player contributes their randomness to the shuffle, ensuring no one party can predict or influence the card order. Once shuffled, cards are encrypted and distributed to players, who can decrypt their own hands using private keys.
- **Public ledger and private hands**: All other game data (e.g., buy-ins, bets, community cards) is recorded publicly on-chain, ensuring transparency. However, player-specific data (private hands, strategies) remains hidden, utilizing a mix of encryption and ZKPs to ensure the asymmetry necessary for poker. And the game will be built on Aztec smart contracts, using its zkRollup technology to ensure scalability. The Aztec contract will manage public components (such as pot size and community cards) while interacting with Noir’s ZKPs to handle private data like player hands.

![noir 1](https://github.com/user-attachments/assets/f08ef487-293a-445b-956d-06bcad60b555)

## Game flow:

1. **Initialization**: Players connect wallets, join a game, and submit their buy-ins to a smart contract. This contract handles the stakes and payouts. MPC nodes are initialized to handle card shuffling.
2. **Shuffling and Dealing**: Each player contributes random data to the MPC for shuffling. Cards are encrypted, and each player receives their hand, which they can decrypt using their private key. The deck remains hidden to all others until necessary.
3. **Betting Rounds**: Players take turns placing bets. ZKPs validate each action, ensuring that bets are within legal limits without exposing hands. The betting process is managed entirely by the smart contract, ensuring fairness and transparency.
4. **Showdown**: Once the final bets are placed, players use ZKPs to prove their hand's strength. This process ensures no card is revealed prematurely while enabling the smart contract to determine the winner.
5. **End of game**: The smart contract distributes winnings based on the showdown results. Players can cash out their chips, which are automatically converted back into cryptocurrency at the set rate.

# **Assumptions**

1. **Trust assumptions**
    - **MPC as a trustless mechanism**: We assume that the MPC protocol operates securely and cannot be compromised by any one party. Each participant contributes random data to ensure the shuffle is fair. This assumption holds because no single player, or even a subset, controls the entire shuffling process.
    - **No trusted third party**: All proof verifications are automated through ZKPs and managed by the smart contract, meaning players need only trust the system itself, not any single entity.
2. **Privacy and security assumptions**
    - **Private hands via encryption**: The integrity of the system relies on the assumption that each player's cards are securely encrypted and cannot be decrypted by anyone else. Noir’s ZKPs further protect players by ensuring that only the necessary parts of the game are revealed at the correct times.
    - **Soundness of ZKPs**: The design assumes that the ZKPs generated by Noir are sound and complete. This ensures that no invalid actions (such as illegal bets) will be accepted, and no valid actions will be rejected.
    - **Resistance to cheating**: The use of MPC and ZKPs prevents cheating by ensuring that no player can alter the shuffle, deal, or game state. Each action is cryptographically enforced, meaning any attempt to manipulate the game will fail to produce a valid proof and will be rejected by the smart contract.
3. **Technical assumptions**
    - **Gas efficiency**: Given that each action involves proof generation and verification, the design assumes that Noir’s ZKP system and Aztec network will manage gas costs effectively. High gas fees could discourage players from betting or folding often.
    - **Scalability of the MPC network**: We assume the MPC network can handle the computational load required for shuffling and dealing, even as the size of committee or the number of players increases. Additionally, MPC participants should have the computational power to efficiently generate zero-knowledge proofs for game actions.

# **Timeline and Deliverables**

1. **Phase 1**: Research & Design (Month 1-2)
- Deliverable: Detailed system architecture for a decentralized poker game using Noir and MPC on Aztec.
- Finalize cryptographic protocols for privacy and fairness.

2. **Phase 2**: Prototype Development (Month 3-4)
- Deliverable: Functional prototype for poker.
- Implement ZKPs (via Noir) for betting and MPC for shuffling. Deploy Aztec smart contracts for game logic.

3. **Phase 3**: Testing & Launch (Month 5-6)
- Deliverable: Testing results and product mainnet launch.
- Test privacy, fairness, and randomness; launch the game with co-marketing efforts.

4. **Phase 4**: Expansion to Other Social Games (Month 7-8)
- Deliverable: Prototype for additional social games and seamless integration of zkSpin gaming engine (e.g., blackjack, sports betting, prediction market, etc.).
- Adapt core architecture to other games while ensuring privacy and real-time experience.

# **Team**
The zkSpin team hails from MIT, with cutting-edge research in parallel computing, and is backed by a16z CSX. The company CPO has spent 20+ years in top tier web2 and web3 gaming studios and have designed AAA tiles like “Call of Duty”. His games have delighted 180m+ players world-wide, generated $5b in revenue and he is previously the head of games at Polygon.
 
1. **Justin Z (Core Contributor)**:
- Parallel Computing Expert from MIT
- Blockchain Security Expert from CertiK
- Ex-Nvidia / Microsoft / Google / Snap

2. **Charnjit B (CPO)**:
- Seasoned Product and Development Executive
- Ex-Polygon / Yuga Labs / Mythical Games / Activision / Electronic Arts 

3. **Wade S (CBO)**:
- Serial Entrepreneur (IDG Capital-backed, successful exit)
- Senior Private Equity Investor ($1B+ investment)
- MIT Sloan Graduate

# **Start Date**
Nov 10th, 2024

**Comment by Savio-Sou** [comment]
Hi @wadesonghang, thank you for submitting the proposal!

Great if a contact channel(s) to the team and links to relevant experience from the team (e.g. open source code) could be supplemented.

**Comment by Savio-Sou** [comment]
Deadline to update proposals is now extended to October 28th, 2024 if it helps buy a bit of extra time to supplement details. Good luck!
https://github.com/orgs/noir-lang/discussions/6289#discussioncomment-11052718
